docs/development/namespaces.md

```diff
@@ -0,0 +1,8 @@
+# Namespaces
+
+ILIAS is using PSR-4 namespaces in its [src directory](../../src/README.md).
+
+You MAY also use namespaces in the `Modules` or `Services` directories. You SHOULD use `ILIAS\ModuleName` or `ILIAS\ServiceName`, e.g. `ILIAS\Course` or `ILIAS\ActiveRecord` in these cases.
+
+* Please be informed that we might introduce a PSR-4 based autoloading for these components in the future, too, and revise our directory structure accordingly. **You MAY only make use of these namespaces if you are willing to support these necessary future refactorings**.
```

Comment:
@alex40724 thank you very much for the proposal! Just one question: Is `ilCtrl` already ready to work with namespaces? (this is really a question, I just not know)

Reply:
@chfsx Thanks, this is an important question.

No, ilCtrl is not ready. I think there are several places that are not ready. Also the ilObject... related factories and any other place that works "generic factory" like.

But we are witnessing several places where namespaces are already used, e.g. https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/Modules/OrgUnit/PublicApi/OrgUnitUserService.php or in Services/FileDelivery. And we would like to allow these approaches by introducing this small guidline.

Reply:
I listed some limitations or bugs with namespace guis in ilCtrl in https://mantis.ilias.de/view.php?id=26181